atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTopologyValidator.java

```diff
@@ -0,0 +1,261 @@
+import org.immutables.value.Value;
+
+public final class CassandraTopologyValidator {
+    private static final SafeLogger log = SafeLoggerFactory.get(CassandraTopologyValidator.class);
+
+    private final CassandraTopologyValidationMetrics metrics;
+
+    public CassandraTopologyValidator(CassandraTopologyValidationMetrics metrics) {
+        this.metrics = metrics;
+    }
+
+    /**
+     * Checks a set of new Cassandra servers against the current Casssandra servers
+     * to ensure their topologies are matching. This is done to prevent user-led split-brain,
+     * which can occur if a user accidentally provided hostnames for two different Cassandra clusters.
+     *
+     * This is done by coming to a consensus on the topology of the pre-existing hosts,
+     * and then subsequently returning any new hosts which do not match the present topology.
+     *
+     * Of course, there is the base case of all hosts will be new. In this case, we simply check that all
+     * new hosts are in consensus.
+     *
+     * Servers that do not have support for the get_host_ids endpoint are always considered consistent,
+     * even if we cannot come to a consensus on the hosts that do support the endpoint.
+     *
+     * Consensus is defined as:
+     * (1) A quorum of nodes (excluding those without `get_host_ids` support) are reachable.
+     * (2) All reachable nodes have the same set of hostIds.
+     * (3) All Cassandra nodes without get_host_ids support are considered to be matching.
+     *
+     * The above should be sufficient to prevent user-led split-brain as:
+     * (1) The initial list of servers validate that they've at least quorum for consensus of topology.
+     * (2) All new hosts added then must match the set of pre-existing hosts topology.
+     *
+     * There does exist an edge case of, two sets of Cassandra clusters being added (3 and 6 respectively).
+     * On initialization, the Cassandra cluster with 6 will be used as the base case if the other 3 nodes
+     * are down, as this will satisfy quorum requirements. However, the cluster of 6 could be the wrong
+     * cluster, which means we're reading/writing from the wrong cluster! However, this then requires we
+     * check all nodes, which then means we cannot handle Cassandra restarts, thus this is the best we can do.
+     *
+     * @param newlyAddedHosts Set of new Cassandra servers you wish to validate.
+     * @param allHosts All Cassandra servers which must include newlyAddedHosts.
+     * @return Set of Cassandra servers which do not match the pre-existing hosts topology. Servers without
+     * the get_host_ids endpoint will never be returned here.
+     */
+    public Set<CassandraServer> getNewHostsWithInconsistentTopologiesAndRetry(
+            Set<CassandraServer> newlyAddedHosts,
+            Map<CassandraServer, CassandraClientPoolingContainer> allHosts,
+            HumanReadableDuration waitTimeBetweenCalls,
+            HumanReadableDuration maxWaitTime) {
```

Comment:
I'm curious how the rest of this is wired, but `HumanReadableDuration` is conventionally used mainly at the configuration level, with standard Java `Duration` within normal code. In this usage it _might_ be OK, but in general it tends to introduce a lot of indirection when calling other Java APIs, so unless there's a strong reason we want this here I'd advise us to stick with convention.

Reply:
Sure no problemo